src/btrfs.cpp

```diff
@@ -0,0 +1,253 @@
+                   << PARTITION_CONF.arg(model);
+    }
+}
+
+void Btrfs::requestAllocation()
+{
+    if (m_currentProcess) {
+        emit allocationReceived(-1, -1);
+        return;
+    } else if (!m_deviceConfiguration.contains(CONF_PARTITION)) {
+        qCritical() << "Cannot get allocation. No partition configured.";
+        emit allocationReceived(-1, -1);
+        return;
+    }
+
+    m_currentProcess = new QProcess;
+    m_currentProcess->setProgram(BTRFS_PATH);
+    QProcessEnvironment env = QProcessEnvironment::systemEnvironment();
+    env.insert("LANG", "C");
+    m_currentProcess->setProcessEnvironment(env);
+    m_currentProcess->setArguments(QStringList() << "filesystem" << "show");
+
+    connect(m_currentProcess, SIGNAL(finished(int,QProcess::ExitStatus)),
+            this, SLOT(slotAllocationFinished(int,QProcess::ExitStatus)));
+    m_currentProcess->start(QProcess::ReadOnly);
+}
+
+void Btrfs::startBalance(int maxUsagePercent)
+{
+    if (m_currentProcess) {
+        emit balanceFinished(false);
+        return;
+    }
+
+    if (!m_deviceConfiguration.contains(CONF_ROOT_MOUNTPOINT)) {
+        qCritical() << "Cannot get root path. No mountpoint configured.";
+        emit balanceFinished(false);
+        return;
+    }
+
+    m_currentProcess = new QProcess;
+    m_currentProcess->setProgram(BTRFS_PATH);
+    QProcessEnvironment env = QProcessEnvironment::systemEnvironment();
+    env.insert("LANG", "C");
+    m_currentProcess->setProcessEnvironment(env);
+    m_currentProcess->setArguments(QStringList()
+                                   << "balance"
+                                   << "start"
+                                   << QString("-dusage=%1").arg(maxUsagePercent)
+                                   << m_deviceConfiguration.value(CONF_ROOT_MOUNTPOINT));
+
+    connect(m_currentProcess, SIGNAL(finished(int,QProcess::ExitStatus)),
+            this, SLOT(slotBalanceFinished(int,QProcess::ExitStatus)));
+    m_currentProcess->start(QProcess::ReadOnly);
+
+    m_currentProgress = 0;
+    m_progressTimer.start(PROGRESS_TIMER_INTERVAL);
+}
+
+int Btrfs::getBalanceProgress()
+{
+    QProcess process;
+    process.setProgram(BTRFS_PATH);
+    QProcessEnvironment env = QProcessEnvironment::systemEnvironment();
+    env.insert("LANG", "C");
+    process.setProcessEnvironment(env);
+    process.setArguments(QStringList()
+                         << "balance"
+                         << "status"
+                         << m_deviceConfiguration.value(CONF_ROOT_MOUNTPOINT));
+    process.start(QProcess::ReadOnly);
+    process.waitForFinished(1000);
+
+    // don't care about the exit code... it is not 0 for success
+    while (!process.atEnd()) {
+        const QString line = process.readLine().trimmed();
+        if (RE_PROGRESS.indexIn(line) != -1) {
+            const QString progressString = RE_PROGRESS.cap(1);
+            int progress = progressString == "nan"
```

Comment:
I'd be a bit more comfortable with brackets to make things clear, even if operator precedence would make that work without:

```
int progress = ((progressString == "nan") ? 0 : progressString.toInt());
```

Then again, QString [returns 0 when the conversation fails](http://doc.qt.io/qt-5/qstring.html#toInt), so might as well just not do that.